### ISSUE TITLE
LTO may cause segfaults

### ISSUE THREAD

**Issue by JD557** [scala-native/scala-native#3172]
See the builds from https://github.com/JD557/minart/pull/322

All builds failed with the segfaults until I disabled LTO on the tests.
Some notes that might be relevant:

- I could not reproduce this problem locally, but it happened consistently on GitHub actions (maybe related to a certain LLVM version?)
- The problematic test loads an AIFF file from a resource, writes the AIFF file to a `ByteArrayOutputStream` and reads it back from a `ByteArrayInputStream`
- There are more tests with a similar pattern (to test other file formats) that worked as expected.

**Comment by LeeTibbert** [comment]
Sorry that you are experiencing this problem and thank you for reporting

Notes:  From a hasty peek at the files in the referenced repository

compile_test.yml --  Ubuntu-latest   -- so unix, and probably Ubuntu 22-04(?). That takes my standard "LTO is
busted on macOS" out of play.                                             

build.sbt:
```
 Test / nativeMode    := "debug",
  nativeLTO            := "thin",
```

The base note mentions running tests.

I am a provincial, but having "LTO=thin" with "debug" looks strange.  What happens if that gets
changed to "release", similar to the other nativeMode?  I do not mean to drag you down a time wasting
rabbit hole.   Agreed, in a better world, the combination would either flag out as incompatible or work.

**Comment by JD557** [comment]
> I am a provincial, but having "LTO=thin" with "debug" looks strange

Yes, that was an oversight on my part, I already changed it on that PR.
I'll try to test it with `release`+`thin` in a bit

**Comment by JD557** [comment]
Looks like that worked as well: https://github.com/JD557/minart/actions/runs/4218614159/jobs/7323250192

(Now I'm wondering if the problem is actually fixed or if it is going to appear again in the future once I update the code)

**Comment by LeeTibbert** [comment]
Thank you for reporting the results.  Sounds like either you or I should open an Issue, referencing this one,
that LTO and release=(None, Debug) should not be allowed together. 

There might be reasons during  development/proving of LTO that an individual SN devo might want to
have them together, but they should know how to hack the source to achieve that .

> (Now I'm wondering if the problem is actually fixed or if it is going to appear again in the future once I update the code)

Ah, the key question. Once trust is lost, it is hard to regain.  Once one has had a cold bath whilst trying
to cross an ice covered lake one constantly asks: now how thin _is_ that ice again?

I am facing the issue of establishing trust in some software in an development effort I am working on.
Is what I am doing for configuration resulting in the changes I want/need?  So far: arrgh!

I believe "thin"+LTO have had two or so years of pretty extensive use by projects larger than
just it passing its own CI. Sounds like you are setting up some tests to see if you get the 
output you expect.  One can not "test in" confidence in correctness, but they sure help.
Reports, like this, from others in the community also help.

 Only major problem I know is that it fails on macOS consistently. I took a run at that two weeks
or so ago and got interrupted before I could establish if that problem still exists or not.

**Comment by ekrich** [comment]
Just as a FYI, the restriction to allow debug + thin was lifted (requested by Lorenzo) whereas previously it was not allowed. Poking through online information suggested that allowing it could be useful for some debugging situations.

**Comment by LeeTibbert** [comment]
ekrich, Thank you for the change history. Classic inclusion/exclusion problem.  No matter what you do, somebody
is going to be displeased.   This has been reported enough times, and little progress has been made on it over
the years, that it seems like it should be hidden behind (one Global 
"ScalaNative_Blade_Guards_OFF_I_will_take_consequences" config option.

Next weeks problem....

**Comment by ekrich** [comment]
@JD557 Could you report your OS, clang version etc? It might be in your references but here would be nicer.

**Comment by JD557** [comment]
I believe the problem happened in Ubuntu 22.04.02, whose default Clang version seems to be 14

Locally I was using with Windows 11 and Clang 12 with no problems (unfortunately, this was a while ago, so it's hard for me to say the precise windows version).
